src/LanguageServer/Protocol/Handler/ServerLifetime/LspServiceLifeCycleManager.cs

```diff
@@ -43,7 +43,7 @@ public async Task ShutdownAsync(string message = "Shutting down")
         // Shutting down is not cancellable.
         var cancellationToken = CancellationToken.None;
 
-        var hostWorkspace = _lspWorkspaceRegistrationService.GetAllRegistrations().SingleOrDefault(w => w.Kind == WorkspaceKind.Host);
+        var hostWorkspace = _lspWorkspaceRegistrationService.GetAllRegistrations().FirstOrDefault(w => w.Kind == WorkspaceKind.Host);
```

Comment:
What needed this change?  I would not generally expect there to be two host workspaces ever.

Reply:
Unfortunately with this refactoring as it is, there are two host workspaces being created. There's the real one, and then the one that CreateTestLspServerAsync is creating from the markups that are being passed into the various helpers. It creates a TestWorkspace which then gets passed around. There is a deeper cleanup needed to make that go away -- a lot of the use of the TestWorkspace is to get workspace services or the MEF export provider. There's also a lot of places where the workspace gets handed around to give it back to a static helper method which is a bit strange in places too.

I don't see any fundamental blockers other than it'll take a bit of time. Unless you are particularly concerned I'll add a comment here just to clarify what's going on.

Reply:
This is also a case where it's a bit strange we're getting a host workspace just to get this other service -- like why is this a workspace service in the first place? What does it mean to be a workspace service?